### PR TITLE
GH-72904: Add `glob.translate()` function

### PR DESCRIPTION
**Pull request by barneygale**
Add `glob.translate()` function that converts a pathname with shell wildcards to a regular expression. The regular expression is used by pathlib to implement `match()` and `glob()`.

This function differs from `fnmatch.translate()` in that wildcards do not match path separators by default, and that a `*` pattern segment matches precisely one path segment. When *recursive* is set to true, `**` pattern segments match any number of path segments, and `**` cannot appear outside its own segment.

In pathlib, this change speeds up directory walking (because `_make_child_relpath()` does less work), makes path objects smaller (they don't need a `_lines` slot), and removes the need for some gnarly code.

* Issue: gh-72904

----
:books: Documentation preview :books:: https://cpython-previews--106703.org.readthedocs.build/

### CODE REVIEW

**Comment by barneygale** [comment]
~20% globbing speedup:

```shell
$ ./python -m timeit -s 'from pathlib import Path; p = Path()' 'list(p.glob("**/*", follow_symlinks=False))'
2 loops, best of 5: 175 msec per loop  # before
2 loops, best of 5: 146 msec per loop  # after
```

File: Doc/library/fnmatch.rst
Comment (jaraco) [inline Doc/library/fnmatch.rst]:
```suggestion
   the separator is used to split the pattern into segments, where:
```

File: Doc/library/fnmatch.rst
Comment (jaraco) [inline Doc/library/fnmatch.rst]: This phrasing is odd. Any position other than what? Maybe you mean:
```suggestion
   - If ``**`` appears in any position other than a full path segment within the pattern,
     :exc:`ValueError` is raised.
```

File: Lib/fnmatch.py
Comment (jaraco) [inline Lib/fnmatch.py]: I'm wondering if instead of entangling these two modes, there should be two functions, like `translate` and `translate_full`, possibly with common behaviors extracted into helpers.

File: Lib/fnmatch.py
Comment (jaraco) [inline Lib/fnmatch.py]: Only allowing for one path character may be limiting in a Windows environments, where both `os.sep` and `os.altsep` are valid path separators. I'm thinking this function should accept any number of separator characters and the default should be `os.sep + (os.altsep if os.altsep else '')`. That is, be portable by default.

File: Lib/fnmatch.py
Comment (jaraco) [inline Lib/fnmatch.py]: Philosophy: I'd avoid the highly nested, stateful code here. It's difficult to decipher and easy to get wrong. I'd much prefer to see something like https://github.com/jaraco/zipp/blob/main/zipp/glob.py, which implements very similar behavior in just three expressions.

That said, I'm okay to accept this imperative implementation, in part because it follows the existing pattern and has tests to cover the intended outputs.

File: Lib/test/test_fnmatch.py
Comment (jaraco) [inline Lib/test/test_fnmatch.py]: TIL: I was surprised by the `?s` (and had to look it up). It allows newline characters to match.

File: Lib/test/test_fnmatch.py
Comment (jaraco) [inline Lib/test/test_fnmatch.py]: I'd love to see some more functional tests capturing the matching of representative filenames with edge cases, similar to [these cases](https://github.com/jaraco/zipp/blob/0dc7e5e24ac2fc10fb95cf5105b1193fff53a27e/tests/test_path.py#L473-L514) or like the functional tests above (but `sep`-based versions).

**Review by jaraco — APPROVED** [review]

File: Lib/fnmatch.py
Comment (jaraco) [inline Lib/fnmatch.py]: Prior to `sep`, `*` translates to `.*` not `.+`:

```
>>> fnmatch.translate('*')
'(?s:.*)\\Z'
```

And that's the behavior I see in shell too:

```
@ ls *README.rst
README.rst
```

I'm thinking `glob('*')` should match zero characters.

Comment (barneygale) [reply Lib/fnmatch.py]: When it appears as its own segment, `*` consumes at least one character in both pathlib and `glob.glob()`:

```python
>>> import os, glob
>>> os.listdir()
['empty']
>>> os.listdir('empty/')
[]
>>> glob.glob('empty/*')
[]
>>> glob.glob('empty/**', recursive=True)
['empty/']
```

This happens in bash 5.0 too:

```
$ ls empty/
$ ls empty/*
ls: cannot access 'empty/*': No such file or directory
```

(note that the shell is matching nothing, so the `empty/*` argument is passed unchanged to `ls`)

File: Lib/fnmatch.py
Comment (barneygale) [reply Lib/fnmatch.py]: That code is really nice :o I'll look at adopting that approach I think!

File: Lib/fnmatch.py
Comment (barneygale) [reply Lib/fnmatch.py]: Maybe `glob.translate()`? Downside there is that it's inconsistent with the defaults for the *recursive* and *include_hidden* arguments to `glob.glob()`.

File: Lib/fnmatch.py
Comment (jaraco) [reply Lib/fnmatch.py]: I'm glad you like it. In https://github.com/jaraco/zipp/pull/108, I've generalized the implementation to align more with some of the assumptions here. Supporting different separators added quite a bit more complexity because that one factor now ties deep into the implementation. I chose to implement a class to store that state, but nested functions or (less elegantly) parameter passing could also have been used. Addressing that one concern, much of the prior simplicity is lost.

Feel free to use that as inspiration, or if you'd like, I can work on contributing that as `glob.translate` (or similar). If we're heading that route, I'd also be willing to work on expanding the unit tests (as I requested of you for this implementation).

File: Lib/fnmatch.py
Comment (jaraco) [reply Lib/fnmatch.py]: Aah, yes. So it's `fnmatch.translate` that's probably missing this expectation. I think I agree the bash/glob behavior is preferable, though it does complicate the implementation.

File: Lib/fnmatch.py
Comment (jaraco) [reply Lib/fnmatch.py]: I like `glob.translate`. I think it's okay if it's not consistent with the interfaces for other functions, as its naming is clearly a departure from any prior expectation. I think a note in the docstring (or docs) would be sufficient to help users avoid confusion.

File: Lib/fnmatch.py
Comment (barneygale) [reply Lib/fnmatch.py]: I've been experimenting with your implementation and I think it's probably the way forward, but I had trouble getting all the pathlib tests passing. Perhaps we could look at adopting it after this PR lands?

**Comment by barneygale** [comment]
I've moved this to a new `glob.translate()` function.

It was easy enough to implement a *recursive* argument, so I did that and made its default `False` to match `glob()`.

It's much harder to implement an *include_hidden* argument, so I've left that for now. I don't feel great about it, tbh.

**Comment by barneygale** [comment]
Right, I've added support for `include_hidden=False` and made it the default (to match `glob.glob()`). Pathlib supplies `include_hidden=True`.

**Comment by barneygale** [comment]
@serhiy-storchaka tagging you as an occasional `glob` contributor - would you be happy for this to land? Thank you!

File: Lib/fnmatch.py
Comment (barneygale) [reply Lib/fnmatch.py]: I've finally got this working using regular expressions. New PR: #109879

**Comment by barneygale** [comment]
Marked as a draft because I'm not satisfied with the implementation - it's too complicated, especially the `include_hidden=False` support.

I've opened #109879, which revamps `fnmatch.translate()` and neatly sets the stage for `glob.translate()`.

**Comment by barneygale** [comment]
Right, after some futzing around I'm going to mark this PR as ready again.

In `fnmatch.py`, I've split the `translate()` method into `_translate()` and `_join_translated_parts()`. This minimises the diff, risk, and performance impact in that module.

(In #109879 I've re-implemented `fnmatch.translate()`, but that PR is an optional side-quest now)

In `glob.py`, I've spent some time making the implementation of `translate(include_hidden=False)` as clear and performant as I can. There's still a fair whack of regex involved but I _think_ it's followable.

In `pathlib.py`, I've made pattern matching use `str(path)` directly, rather than a slight variant that represents empty paths as `''` rather than `'.'`.

**Comment by barneygale** [comment]
Timings:

```
$ ./python -m timeit -n 5 -s 'from glob import glob' 'list(glob("**/*", recursive=True, include_hidden=True))'
5 loops, best of 5: 105 msec per loop  # for interest
$ ./python -m timeit -n 5 -s 'from pathlib import Path; p = Path()' 'list(p.glob("**/*", follow_symlinks=True))'
5 loops, best of 5: 86.6 msec per loop  # before
5 loops, best of 5: 72.7 msec per loop  # after
```

So `Path.glob()` is ~20% faster than before, and ~45% faster than `glob.glob()`

File: Lib/glob.py
Comment (AA-Turner) [inline Lib/glob.py]: Tiny memory improvement
```suggestion
            seps = (os.path.sep, os.path.altsep)
```

File: Lib/glob.py
Comment (AA-Turner) [inline Lib/glob.py]:
```suggestion
    escaped_seps = ''.join(map(re.escape, seps))
```

File: Lib/glob.py
Comment (AA-Turner) [inline Lib/glob.py]:
```suggestion
        any_segments = f'(?:{one_segment})*'
        any_last_segments = f'{any_segments}(?:{one_last_segment})?'
```

File: Lib/glob.py
Comment (AA-Turner) [inline Lib/glob.py:278]: `one_segment` could maybe be factored out as it's the same in either case?

File: Lib/pathlib.py
Comment (AA-Turner) [inline Lib/pathlib.py]: Could you add an explanatory comment about `(\.\Z)?+`?

File: Lib/glob.py
Comment (AA-Turner) [inline Lib/glob.py:311]: Why ``\Z`` over ``$``?

**Review by AA-Turner — COMMENTED** [review]
A desk review:

A

File: Lib/glob.py
Comment (barneygale) [reply Lib/glob.py:311]: `$` can match before a newlines in [`re.MULTILINE`](https://docs.python.org/3/library/re.html#re.MULTILINE) mode, and newlines are valid path characters on POSIX.

`fnmatch.translate()` uses the same incantation.

File: Lib/glob.py
Comment (barneygale) [reply Lib/glob.py:278]: In the `else:` block, it's used to define `any_segments`.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: Done! Hopefully it makes sense?

File: Lib/glob.py
Comment (barneygale) [reply Lib/glob.py:278]: I played around with refactoring this, but I couldn't find a way to make the code clearer.

**Comment by barneygale** [comment]
I've realised that the docs are a bit skew-whiff. Fix is in a separate PR: #110418

**Comment by encukou** [comment]
Is it correct to keep duplicated path separators?
```python
>>> glob.translate('a//b')
'(?s:a//b)\\Z'
```

**Comment by barneygale** [comment]
> Is it correct to keep duplicated path separators?
> 
> ```python
> >>> glob.translate('a//b')
> '(?s:a//b)\\Z'
> ```

`glob()` keeps them:

```python
>>> import os
>>> os.makedirs('a/b')
>>> glob.glob('a//b')
['a//b']
```

So I reckon yes?

**Comment by barneygale** [comment]
Also, the number of additional slashes is meaningful in some cases, e.g. in Windows UNC paths or POSIX paths starting with two forward slashes. I don't think a pattern like `/foo` should match a path like `//foo`, and vice-versa.

**Comment by barneygale** [comment]
Hey @encukou, do you think I can merge this, or should I wait for a more complete review?

**Comment by encukou** [comment]
Oh, I should have been more clear that I wouldn't get to a thorough review in any reasonable time.
You already have a few approvals, merge away!

**Pull request merged**
